src/permutation.jl

```diff
@@ -0,0 +1,65 @@
+export PermutationTest
+
+function ptstats(x,y,f)
+    xy = vcat(x,y)
+    rx = 1 : length(x)
+    ry = length(xy)-length(y)+1 : length(xy)
+    T = typeof(f(xy))  # is there a better way?  e.g., does a funtype() exist, like eltype()??
+    (xy, rx, ry, T)
+end
+
+immutable PermutationTest <: HypothesisTest
+    observation
+    samples
+end
+
+"""
+    PermutationTest(x::Vector, y::Vector, f::Function [, n::Int])
+
+Perform a permutation test (a.k.a. randomization test) of the null hypothesis
+that the probability of `f(x)` is equal to `f(y)`.
+
+If `n` is *not* supplied, all possible permutations are sampled.  Otherwise,
+only `n` of the `factorial(length(x)+length(y))` permutations are sampled.
+"""
+function PermutationTest{R<:Real,S<:Real}(x::AbstractVector{R}, y::AbstractVector{S}, f::Function)
+    xy, rx, ry, T = ptstats(x,y,f)
+    P = permutations(xy)
+    xsamples = Vector{T}(length(P))
+    ysamples = Vector{T}(length(P))
+    for (i,p) in enumerate(P)
+        xsamples[i] = f(view(p,rx))
+        ysamples[i] = f(view(p,ry))
+    end
+    PermutationTest(f(x) - f(y), xsamples - ysamples)
+end
+
+function PermutationTest{R<:Real,S<:Real}(x::AbstractVector{R}, y::AbstractVector{S}, f::Function, n::Int)
+    xy, rx, ry, T = ptstats(x,y,f)
+    xsamples = Vector{T}(n)
+    ysamples = Vector{T}(n)
+    for i = 1:n
+        shuffle!(xy)
+        xsamples[i] = f(view(xy,rx))
+        ysamples[i] = f(view(xy,ry))
+    end
+    PermutationTest(f(x) - f(y), xsamples - ysamples)
+end
+
+function pvalue(apt::PermutationTest; tail=:both)
+    if tail == :both
+        count = sum(abs(apt.observation) .>= @compat abs.(apt.samples))
```

Comment:
No longer necessary to `@compat` this, since support for 0.4 has been dropped (and 0.5 as well)